modules/javafx.graphics/src/main/java/com/sun/javafx/tk/quantum/QuantumToolkit.java

```diff
@@ -1497,44 +1559,58 @@ public void run() {
-                        if (pImage.rt.readPixels(ib, pImage.rt.getContentX(),
-                                pImage.rt.getContentY(), w, h))
-                        {
-                            pImage.setImage(com.sun.prism.Image.fromIntArgbPreData(ib, w, h));
-                        } else {
-                            pImage.dispose();
-                            pImage = null;
+                            (QuantumImage) params.platformImage : new QuantumImage((com.sun.prism.Image) null);
+
+                    int maxTextureSize = rf.getMaximumTextureSize();
+                    if (h > maxTextureSize || w > maxTextureSize) {
+                        // The requested size for the screenshot is too big to fit a single texture,
+                        // so we need to take several snapshot tiles and merge them into pImage
+                        if (pImage.image == null) {
+                            pImage.setImage(com.sun.prism.Image.fromIntArgbPreData(IntBuffer.allocate(w * h), w, h));
+                        }
+                        // Find out it is is possible to divide up the image in tiles of the same size
+                        AtomicBoolean exactWidthDivFound = new AtomicBoolean(false);
+                        int tileWidth = computeOptimumTileSize(w, maxTextureSize, exactWidthDivFound);
+                        AtomicBoolean exactHeightDivFound = new AtomicBoolean(false);
+                        int tileHeight = computeOptimumTileSize(h, maxTextureSize, exactHeightDivFound);
+                        IntBuffer buffer = IntBuffer.allocate(tileWidth * tileHeight);
+                        // choose tiling loop order to minimize surface thrashing
+                        if (exactHeightDivFound.get()) {
+                            for (int xOffset = 0; xOffset < w; xOffset += tileWidth) {
+                                tileWidth = Math.min(tileWidth, w - xOffset);
+                                for (int yOffset = 0; yOffset < h; yOffset += tileHeight) {
+                                    tileHeight = Math.min(tileHeight, h - yOffset);
+                                    renderTile(x, xOffset, y, yOffset, tileWidth, tileHeight,
+                                            buffer, rf, tileRttCache, pImage);
+                                }
+                            }
+                        }
+                        else {
+                            for (int yOffset = 0; yOffset < h; yOffset += tileHeight) {
+                                tileHeight = Math.min(tileHeight, h - yOffset);
+                                for (int xOffset = 0; xOffset < w; xOffset += tileWidth) {
+                                    tileWidth = Math.min(tileWidth, w - xOffset);
+                                    renderTile(x, xOffset, y, yOffset, tileWidth, tileHeight,
+                                            buffer, rf, tileRttCache, pImage);
+                                }
+                            }
```

Comment:
The `if (exactHeightDivFound.get())` and `else` block here look effectively same. Could you please take a re-look or explain what am I missing here.

Reply:
Functionally, both side of the condition actually do the exact same thing, but the side effects of which dimension is processed in the inner loop sometime have measurable consequences in terms of performances; I've added the following comment to the code to better explain this:
```
  // In order to minimize the number of time we have to resize the underlying
  // surface for capturing a tile, choose a dimension that has an exact divider
  // (if any) to be processed in the inner most loop.
  // E.g. looping on width then height in the example bellow requires four
  // surface resizing, whereas the opposite requires only two:
  //
  //       for (w;;)                    for (h;;)
  //           for(h;;)                     for(w;;)
  //       -----------------           -----------------
  //       |       |       |           |       |       |
  //       |   1   |   3   |           |   1   |   2   |
  //    h  |       |       |        h  |       |       |
  //       -----------------           -----------------
  //       |   2   |   4   |           |   3   |   4   |
  //       -----------------           -----------------
  //               w                           w
```

Reply:
Than for the additional comment, this will reduce the number of textures created by below call in `renderTile()`
`com.sun.prism.RTTexture rt = tileImg.getRT(w, h, rf);`